### PR TITLE
test/extended: refactor external DNS test

### PR DESCRIPTION
**Pull request by ironcladlou**
Replace the "wildcard DNS" ingress test with a more general one which tests
external connectivity to well-known core cluster services.

/cc @openshift/sig-network-edge @smarterclayton

### CODE REVIEW

**Comment by ironcladlou** [comment]
The scenarios could be parallelized

File: test/extended/operators/routable.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+			}
+			o.Expect(err).NotTo(o.HaveOccurred(), "error getting default router service: %v", err)
+		}
+
+		if svc.Spec.Type != corev1.ServiceTypeLoadBalancer {
+			g.Skip("default router is not exposed by a load balancer service")
+		}
+	})
+
+	g.It("expose cluster services outside the cluster", func() {
```
Comment (Miciah) [inline test/extended/operators/routable.go]: Missing "should".

File: test/extended/operators/routable.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+				if len(route.Status.Ingress) == 0 || len(route.Status.Ingress[0].Host) == 0 {
+					return false, nil
+				}
+				hostname = route.Status.Ingress[0].Host
+				return true, nil
+			})
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			url := fmt.Sprintf("%s://%s", test.scheme, hostname)
+			g.By(fmt.Sprintf("verifying the %s/%s route serves 200 from %s", test.routeNamespace, test.routeName, url))
```
Comment (Miciah) [inline test/extended/operators/routable.go]: It's not necessarily "200"; need to interpolate `test.expect`.

File: test/extended/operators/routable.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+			err = waitForURLOK(ns, execPodName, url, test.expect, endpointWait)
+			o.Expect(err).NotTo(o.HaveOccurred())
+		}
+	})
+})
+
+func waitForURLOK(ns, execPodName, url, expect string, timeout time.Duration) error {
```
Comment (smarterclayton) [inline test/extended/operators/routable.go]: This code is fine, but you might want to consider using the route tester harness

	"github.com/openshift/origin/test/extended/util/url"

Lets you do some more efficient testing:

```
			t.Within(
				3*time.Minute,
				url.Expect("GET", "http://1.ingress-test.com/test").Through(host).HasStatusCode(200),
				url.Expect("GET", "http://1.ingress-test.com/other/deep").Through(host).HasStatusCode(200),
				url.Expect("GET", "http://1.ingress-test.com/").Through(host).HasStatusCode(503),
				url.Expect("GET", "http://2.ingress-test.com/").Through(host).HasStatusCode(200),
				url.Expect("GET", "https://3.ingress-test.com/").Through(host).SkipTLSVerification().HasStatusCode(200),
				url.Expect("GET", "http://3.ingress-test.com/").Through(host).RedirectsTo("https://3.ingress-test.com/", http.StatusFound),
			)
```

Comment (ironcladlou) [reply test/extended/operators/routable.go]: Great tip, thanks

**Comment by ironcladlou** [comment]
Locally I'm getting some weird errors that don't seem related to the (successful) curl calls:

```
Feb 18 11:30:53.513: INFO: Running '/usr/local/bin/kubectl --server=https://api.dmace.devcluster.openshift.com:6443 --kubeconfig=/Users/dmace/Projects/installer/clusters/openshift-dev-dmace/auth/kubeconfig exec --namespace=e2e-test-operators-routable-v5rr9 execpod -- /bin/sh -c set -euo pipefail
function json_escape() {
  python -c 'import json,sys; print json.dumps(sys.stdin.read())'
}
# Test: 0
rc=0
curl -X GET   -s -S -o /tmp/body -D /tmp/headers "https://console-openshift-console.apps.dmace.devcluster.openshift.com" -w '{"code":%{http_code}}' -k 2>/tmp/error 1>/tmp/output || rc=$?                     
echo "{\"test\":0,\"rc\":$(echo $rc),\"curl\":$(cat /tmp/output),\"error\":$(cat /tmp/error | json_escape),\"body\":\"$(cat /tmp/body | base64 -w 0 -)\",\"headers\":$(cat /tmp/headers | json_escape)}"       
# Test: 1
rc=0
curl -X GET   -s -S -o /tmp/body -D /tmp/headers "https://prometheus-k8s-openshift-monitoring.apps.dmace.devcluster.openshift.com" -w '{"code":%{http_code}}' -k 2>/tmp/error 1>/tmp/output || rc=$?           
echo "{\"test\":1,\"rc\":$(echo $rc),\"curl\":$(cat /tmp/output),\"error\":$(cat /tmp/error | json_escape),\"body\":\"$(cat /tmp/body | base64 -w 0 -)\",\"headers\":$(cat /tmp/headers | json_escape)}"'      
Feb 18 11:30:54.497: INFO: stderr: "/bin/sh: python: not found\nbase64: unrecognized option: w\nBusyBox v1.26.2 (2018-05-30 13:51:20 GMT) multi-call binary.\n\nUsage: base64 [-d] [FILE]\n\nBase64 encode or decode FILE to standard output\n\t-d\tDecode data\n/bin/sh: python: not found\n/bin/sh: python: not found\nbase64: unrecognized option: w\nBusyBox v1.26.2 (2018-05-30 13:51:20 GMT) multi-call binary.\n\nUsage:
base64 [-d] [FILE]\n\nBase64 encode or decode FILE to standard output\n\t-d\tDecode data\n/bin/sh: python: not found\n"                                                                                        
Feb 18 11:30:54.497: INFO: stdout: "{\"test\":0,\"rc\":0,\"curl\":{\"code\":200},\"error\":,\"body\":\"\",\"headers\":}\n{\"test\":1,\"rc\":0,\"curl\":{\"code\":403},\"error\":,\"body\":\"\",\"headers\":}\n"
[AfterEach] [Feature:Platform] Managed cluster should
  /Users/dmace/Projects/origin/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/test/extended/util/cli.go:78                                                                   
STEP: Collecting events from namespace "e2e-test-operators-routable-v5rr9".
STEP: Found 5 events.
Feb 18 11:30:54.573: INFO: At 2019-02-18 11:30:47 -0500 EST - event for execpod: {default-scheduler } Scheduled: Successfully assigned e2e-test-operators-routable-v5rr9/execpod to ip-10-0-155-13.us-east-2.compute.internal
Feb 18 11:30:54.573: INFO: At 2019-02-18 11:30:47 -0500 EST - event for execpod: {kubelet ip-10-0-155-13.us-east-2.compute.internal} Pulling: pulling image "gcr.io/kubernetes-e2e-test-images/hostexec:1.1"   
Feb 18 11:30:54.573: INFO: At 2019-02-18 11:30:49 -0500 EST - event for execpod: {kubelet ip-10-0-155-13.us-east-2.compute.internal} Pulled: Successfully pulled image "gcr.io/kubernetes-e2e-test-images/hostexec:1.1"
Feb 18 11:30:54.573: INFO: At 2019-02-18 11:30:50 -0500 EST - event for execpod: {kubelet ip-10-0-155-13.us-east-2.compute.internal} Created: Created container                                                
Feb 18 11:30:54.573: INFO: At 2019-02-18 11:30:50 -0500 EST - event for execpod: {kubelet ip-10-0-155-13.us-east-2.compute.internal} Started: Started container                                                
Feb 18 11:30:54.671: INFO: skipping dumping cluster info - cluster too large
STEP: Deleting namespaces
Feb 18 11:30:54.744: INFO: namespace : e2e-test-operators-routable-v5rr9 api call to delete is complete                                                                                                        
STEP: Waiting for namespaces to vanish
[AfterEach] [Feature:Platform] Managed cluster should
  /Users/dmace/Projects/origin/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/test/e2e/framework/framework.go:148                                   
Feb 18 11:31:04.817: INFO: Waiting up to 3m0s for all (but 100) nodes to be ready
Feb 18 11:31:04.888: INFO: Running AfterSuite actions on all node
Feb 18 11:31:04.888: INFO: Running AfterSuite actions on node 1
fail [github.com/openshift/origin/test/extended/util/url/url.go:82]: Expected error:
    <*errors.errorString | 0xc421b40ae0>: {
        s: "response 0 could not be decoded: invalid character ',' looking for beginning of value",
    }
    response 0 could not be decoded: invalid character ',' looking for beginning of value
not to have occurred
```

**Comment by ironcladlou** [comment]
Looks like we got a new payload

/retest

**Comment by ironcladlou** [comment]
Looks like EC2 issues.

/retest

**Comment by pravisankar** [comment]
Looks like genuine issue:
```
openshift-tests [Feature:Platform] Managed cluster should should expose cluster services outside the cluster [Suite:openshift/conformance/parallel] 28s
go run hack/e2e.go -v -test --test_args='--ginkgo.focus=openshift\-tests\s\[Feature\:Platform\]\sManaged\scluster\sshould\sshould\sexpose\scluster\sservices\soutside\sthe\scluster\s\[Suite\:openshift\/conformance\/parallel\]$'
fail [github.com/openshift/origin/test/extended/util/url/url.go:82]: Expected error:
    <*errors.errorString | 0xc422ebb480>: {
        s: "response 0 could not be decoded: invalid character ',' looking for beginning of value",
    }
    response 0 could not be decoded: invalid character ',' looking for beginning of value
not to have occurred
```

**Comment by Miciah** [comment]
The image appears to be missing Python.  We need to track that down.

**Comment by ironcladlou** [comment]
Ugh, the problem was right in front of us. When I refactored to `util/url` (which creates its own exec pod), I forgot to remove this from the top of the test function:

```go
execPodName := exutil.CreateExecPodOrFail(oc.AdminKubeClient().Core(), ns, "execpod")
				defer func() { oc.AdminKubeClient().Core().Pods(ns).Delete(execPodName, metav1.NewDeleteOptions(1)) }()
```

Which made an incompatible exec pod with the _same name_ as the one `util/url` wanted to create. The shadowing execpod got re-used. Removing the creation of the old exec pod fixed it.

🤦‍♀️

**Comment by ironcladlou** [comment]
/retest

@smarterclayton tag?

**Comment by smarterclayton** [comment]
Your latest commit hasn’t gone green, have you reviewed every retest to make sure it hasn’t continued to fail?

**Comment by ironcladlou** [comment]
I think the past 6 of 7 failures have been cluster provisioning failures, and the other 1 was metrics.

**Comment by ironcladlou** [comment]
More AWS issues.

/retest

**Pull request merged**